### PR TITLE
Add BIP324 transport support and session_id for identity binding

### LINKED ISSUE #801: Encrypted Coinswap Transport with BIP324
# Project: Encrypted Coinswap Transport with BIP324

## Summary
Coinswap currently exchanges maker/taker protocol messages over plaintext length-prefixed CBOR on top of `TcpStream`. Tor helps with network privacy, but the application payload itself should still be protected with an authenticated encrypted transport.

This project is to design and integrate a BIP324-based transport layer for Coinswap using the [`rust-bitcoin/bip324`](https://github.com/rust-bitcoin/bip324) library, so that maker/taker communication is encrypted by default, authenticated at the peer-identity level, and usable by both legacy and taproot flows.

The project should also extend privacy protections to Nostr-based market discovery by routing that communication over Tor and pushing the system toward a no-clearnet architecture wherever practical.

The goal is not to replace the Coinswap protocol itself. The goal is to place a secure transport layer underneath the existing message exchange.

## Problem
Today, protocol messages can leak useful metadata if observed in transit, and plaintext transport leaves too much trust in the surrounding network setup. Coinswap needs a transport layer that:

- protects message confidentiality and integrity
- authenticates the remote peer identity in v1
- works for both maker offer discovery and direct maker/taker communication
- supports both legacy and taproot protocols through one shared transport layer
- does not allow plaintext fallback in production
- reduces or removes clearnet communication from the system to improve privacy

Coinswap also needs its market-discovery path to match the same privacy goals. Nostr-based discovery should not remain a clearnet side channel if the rest of the system is being hardened.

## Expected Outcome
By the end of the project, Coinswap should have a transport layer that:

- uses primitives from [`rust-bitcoin/bip324`](https://github.com/rust-bitcoin/bip324) as the basis for encrypted communication
- is the standard default for maker/taker traffic
- authenticates the remote peer against an advertised or otherwise trusted identity
- can be integrated without rewriting the existing business protocol
- is covered by focused tests and basic migration/config support
- routes Nostr-based market discovery over Tor as part of the privacy model
- minimizes clearnet communication across the system, with the aim of removing it from normal operation

## Recommended Approach
The recommended direction is to introduce a transport abstraction first, then integrate encrypted transport beneath the current CBOR message flow.

At a high level:

1. Isolate transport concerns from protocol logic.
2. Add an encrypted transport implementation backed by [`rust-bitcoin/bip324`](https://github.com/rust-bitcoin/bip324).
3. Ensure the application handshake happens only after the secure channel is established.
4. Reuse the same transport layer across both legacy and taproot paths.
5. Extend the design to maker offer discovery, since offer traffic also leaks metadata.
6. Route Nostr-based market discovery through Tor and avoid relying on clearnet relay access in normal operation.

The contributor should evaluate the available [`rust-bitcoin/bip324`](https://github.com/rust-bitcoin/bip324) crate surface and determine the cleanest integration strategy. The exact module layout, handshake wiring, framing strategy, identity-binding mechanism, and Nostr-over-Tor relay strategy should be designed during implementation rather than copied from a fixed recipe.

## Scope
In scope:

- maker/taker encrypted transport
- peer identity authentication in v1
- transport support for both legacy and taproot flows
- encrypted maker offer discovery connections
- Tor-routed Nostr-based market discovery
- relay-selection strategy for privacy-preserving Nostr communication over Tor
- config and migration behavior needed to make encrypted transport the default
- tests for successful operation and failure cases

Out of scope for the initial project:

- wallet-at-rest encryption
- redesigning the Coinswap application protocol
- broad async/networking refactors unrelated to transport
- maker RPC transport unless there is a strong reason to include it later

## Key Requirements
The implementation should satisfy these project-level requirements:

- Encryption must be the default standard, not an optional extra.
- Plaintext fallback must not be allowed in production.
- Peer identity authentication is required in v1.
- One common transport layer should support both protocol families.
- Existing application-level message types and state machines should change as little as possible.
- Transport failures should fail closed where security policy requires it.
- Nostr-based market discovery should use Tor in normal operation.
- The system should move toward eliminating clearnet communication from normal operation.

## Major Things To Keep In Mind
- Keep the transport layer separate from business logic. The swap protocol should not need major redesign just to gain encrypted transport.
- Preserve compatibility across legacy and taproot flows through one shared interface, not two unrelated implementations.
- Treat Tor as complementary privacy infrastructure, not as a substitute for authenticated encryption.
- Offer discovery matters too. Metadata protection is not only about the later swap-session messages.
- Nostr market discovery is part of the privacy surface and should be treated as such.
- Relay choice matters. The implementer should identify Tor-reachable relays that are good enough for the project’s privacy goals, with preference for ephemeral relays that minimize retention and do not keep logs where possible.
- Identity binding is part of the project, not a future nice-to-have.
- Be careful about accidental plaintext paths, mixed-mode behavior, and downgrade opportunities.
- Avoid leaking sensitive metadata before encryption starts.
- Keep the design maintainable. Security-sensitive code should be isolated, understandable, and testable.

## Deliverables
- A concise design note describing the chosen [`rust-bitcoin/bip324`](https://github.com/rust-bitcoin/bip324) integration strategy
- Encrypted transport integrated into the relevant Coinswap networking paths
- Tor-based Nostr market-discovery integration, including relay-selection rationale
- Configuration behavior that makes secure transport the normal/default mode
- Tests covering happy-path operation, mismatch/failure cases, and regression risk
- Documentation for operators or developers who need to understand the new transport behavior

## Evaluation Criteria
Work on this project should be judged primarily on:

- correctness and safety of the transport design
- clean integration with the existing codebase
- how well peer authentication is handled
- resistance to downgrade or accidental plaintext behavior
- how effectively the design reduces or removes clearnet communication
- how well Nostr-over-Tor discovery fits the overall privacy model
- quality of tests and failure-mode handling
- clarity of the final documentation

## Open Technical Question
One important technical question remains for the implementer to resolve during the project:

- Which parts of the [`rust-bitcoin/bip324`](https://github.com/rust-bitcoin/bip324) crate are stable and usable enough to build the transport cleanly without unnecessary forking or excessive coupling?

This is intentionally left open because exploring the crate surface and selecting the right integration path is part of the project.

### PR DESCRIPTION
**Pull request by YoganshSharma**
# Pull Request

## Description
- saves the corresponding session_id in ConnectionState
-  Replaced raw read_message() / send_message() helpers  with a Bip324Stream wrapper
- Identity binding is done when Maker sends taker AckSwap
  - Taker verifies:  tweakable_point matches the cached offer,  session_id matches the BIP324 transport layer, session_id_sig is valid against the tweakable point
- New Error types `Bip324Error`
 

## Related Issue(s)
Closes #801 

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor / performance improvement
- [ ] Documentation update only
- [ ] CI / Docker / Build changes
- [ ] Other (please describe):

## Protocol Version(s) Affected
- [ ] Legacy (ECDSA — `messages.rs`, `contract.rs`, `handlers.rs`)
- [ ] Taproot-Musig2 (`messages2.rs`, `contract2.rs`, `handlers2.rs`)
- [ ] Both
- [ ] Neither (infrastructure / tooling only)

> **Note:** When modifying protocol logic, changes must be made to **both** versions unless the
> scope is explicitly version-specific.

## Affected Component(s)
- [ ] **makerd** (background daemon)
- [ ] **taker** (CLI client)
- [ ] **maker-cli** (command-line tool)
- [ ] Core protocol / cryptography / library
- [ ] **watch_tower** (contract monitoring & recovery)
- [ ] Docker / deployment scripts
- [ ] Tests / test framework
- [ ] Documentation (`docs/`)
- [ ] Other (please specify):

### Code Quality
- [ ] I ran `cargo +nightly fmt --all` and committed the result
- [ ] I ran `cargo +stable clippy --all-features --lib --bins --tests -- -D warnings` with zero warnings
- [ ] I ran `cargo +stable clippy --examples -- -D warnings` with zero warnings
- [ ] I ran `RUSTDOCFLAGS="-D warnings" cargo +nightly doc --all-features --document-private-items --no-deps` with zero warnings
- [ ] Pre-commit git hook passes (`ln -s ../../git_hooks/pre-commit .git/hooks/pre-commit` if not already set)

### Testing
- [ ] All unit tests pass (`cargo test`)
- [ ] Integration tests pass (`cargo test --features integration-test`)
- [ ] Changes were manually tested on **regtest**
- [ ] End-to-end maker ↔ taker swap tested (where applicable)
- [ ] Test-only code is gated behind `#[cfg(feature = "integration-test")]`

### Documentation
- [ ] Relevant files in the `docs/` folder were updated
- [ ] Complex logic is commented

### Security & Privacy (Critical)
- [ ] This change preserves **trustlessness** and **atomic swap guarantees**
- [ ] No regression in **sybil resistance** or **fidelity bond** logic
- [ ] Tor anonymity and P2P message flow were reviewed
- [ ] No new attack vectors or trust assumptions introduced
- [ ] Edge cases and error handling considered
- [ ] ZMQ subscription integrity (raw tx/block feed) is unaffected
- [ ] `integration-test` feature flag is not reachable in production code paths

## How to Test

```bash
# Standard swap (good baseline)
cargo test --test standard_swap --features integration-test -- --nocapture

# Abort scenarios
cargo test --test abort1 --features integration-test -- --nocapture
cargo test --test taproot_taker_abort1 --features integration-test -- --nocapture

# Malicious behavior & recovery
cargo test --test malice1 --features integration-test -- --nocapture
cargo test --test taproot_timelock_recovery --features integration-test -- --nocapture

# Or run the full suite
cargo test --features integration-test
```

## Summary by CodeRabbit

* **New Features**
  * Encrypts maker↔taker communication with BIP324 streams across swap, legacy, and offer synchronization flows.
  * Maker acknowledgement now includes a BIP324 session id plus signature; takers validate these during negotiation.
  * Offer synchronization now uses the configured Bitcoin network when establishing maker sessions.
* **Bug Fixes**
  * Improved handling and reporting of BIP324 protocol/connection-close failures for clearer network error behavior.
* **Chores**
  * Updated the pinned BIP324 dependency revision.

### CODE REVIEW

File: src/protocol/common_messages.rs [hunk 1]
```diff
@@ -99,22 +99,39 @@ pub struct SwapDetails {
...
 }
 
+/// Acknowledgment of swap details from Maker.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum AckSwap {
+    /// Maker accepted the swap and provides the tweakable point, session ID, and session ID signature.
+    Acknowledged(AckSwapDetails),
+    /// Maker rejected the swap.
+    Rejected,
+}
```
Comment (mojoX911) [inline src/protocol/common_messages.rs]: This wrapping seems redundant. We can as well use the old structure of `AckSwapDetails` and add the new fields there. Also the `Rejected` message is unused. In no place is the maker actually emitting this message. But the taker is handling for it, which is basically dead code. 

So the whole thing can be consolidated into one single struct `AckSwapDetails`.

File: src/maker/handlers.rs [hunk 2]
```diff
@@ -478,7 +480,14 @@ fn handle_swap_details<M: Maker>(
 
     maker.store_connection_state(&details.id, state)?;
 
-    let (_, tweakable_point) = maker.get_tweakable_keypair()?;
+    let (tweakable_privkey, tweakable_point) = maker.get_tweakable_keypair()?;
+
+    let session_id = state.session_id.unwrap();
```
Comment (mojoX911) [inline src/maker/handlers.rs]: Handle this unwrap. Should not hard panic here. If the session id is not set, which should not be the case, set the session id here itself. Add comment that session id should always be aviable at this stage.

File: src/maker/server.rs [hunk 3]
```diff
@@ -12,13 +12,64 @@ use std::{
...
 };
 
+use bip324::io::Payload;
+
+struct MakerBip324Wrapper {
+    protocol: bip324::io::Protocol<TcpStream, TcpStream>,
+}
+
+impl MakerBip324Wrapper {
+    fn new(stream: TcpStream, maker: Arc<MakerServer>) -> Result<Self, MakerError> {
```
Comment (mojoX911) [inline src/maker/server.rs]: We don't really need to pass an Arc<Maker>. We just need the network. So pass in the network value only.

File: src/taker/api.rs [hunk 4]
```diff
@@ -364,6 +366,53 @@ impl MakerConnection {
...
+
+impl TakerBip324Wrapper {
+    pub fn new(stream: TcpStream, network: bitcoin::Network) -> Result<Self, ProtocolError> {
+        let reader = stream.try_clone()?;
+        let writer = stream;
+        let protocol = bip324::io::Protocol::new(
+            network.magic(),
+            bip324::Role::Initiator,
+            None,
+            None, // no garbage or decoys
```
Comment (mojoX911) [inline src/taker/api.rs]: This and the maker server's wrapper are exactly the same. So it can be generalized into a single structure and placed inside utill.rs. Use the common struct for both takers and makers.

File: src/taker/api.rs [hunk 5]
```diff
@@ -364,6 +366,53 @@ impl MakerConnection {
     }
 }
 
+pub(crate) struct TakerBip324Wrapper {
+    protocol: bip324::io::Protocol<TcpStream, TcpStream>,
+}
```
Comment (mojoX911) [inline src/taker/api.rs]: Naming is off. Do something like this. Make this the common struct `Bip324Stream`.  Add doc comment explaining what this is. Move it to utill.rs.


```suggestion
pub(crate) struct Bip324Stream {
    stream: bip324::io::Protocol<TcpStream, TcpStream>,
}
```

File: src/taker/api.rs [hunk 6]
```diff
@@ -1369,16 +1435,28 @@ impl Taker {
                         ));
                     }
 
+                    let maker_session_id = ack_details.session_id;
+                    let taker_session_id = stream.protocol.session_id();
+                    if &maker_session_id != taker_session_id {
+                        return Err(TakerError::Net(Bip324Error::SessionIdMismatch.into()));
```
Comment (mojoX911) [inline src/taker/api.rs]: This can be done in one line.

File: src/taker/api.rs [hunk 7]
```diff
@@ -1704,7 +1782,10 @@ impl Taker {
             .and_then(|_| socket.set_write_timeout(Some(timeout)))
             .map_err(|e| TakerError::General(format!("Failed to set socket timeout: {}", e)))?;
 
-        Ok(socket)
+        let encrypted_protocol =
```
Comment (mojoX911) [inline src/taker/api.rs]: It's a stream structure, not a protocol. Protocol is a misnomer here.  

```suggestion
        let bip324_stream =
```

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Mostly looks good. Few comments and change suggestions.

**Comment by mojoX911** [comment]
@YoganshSharma pls fix the linter and mark this as ready for review..

**Comment by mojoX911** [comment]
@YoganshSharma some networking code has changed due to recent merges. They will impact changes here too. Please rebase this PR.

File: src/utill.rs [hunk 8]
```diff
@@ -969,6 +970,57 @@ pub fn interactive_select(
...
+            None, // no garbage or decoys
+            reader,
+            writer,
+        )?;
+
+        Ok(Self { protocol })
+    }
+
+    /// Reads the next genuine application message from the BIP324 stream.
+    pub(crate) fn read_message(&mut self) -> Result<Vec<u8>, NetError> {
```
Comment (Godzilla-oss) [inline src/utill.rs]: Could `read_message` deserialize internally and return the typed message? No caller uses the raw bytes

Comment (YoganshSharma) [reply src/utill.rs]: Sure I can try using generic for this

**Comment by Godzilla-oss** [comment]
@YoganshSharma Can you also check if we can use this here if the performance gain is reasonable, as mentioned in the BIP324 create: "For optimal performance, wrap your `reader` in a [`std::io::BufReader`]. The protocol makes many small reads during handshake and operation."

File: src/protocol/common_messages.rs [hunk 9]
```diff
@@ -104,17 +104,25 @@ pub struct SwapDetails {
...
 #[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
 pub struct AckSwapDetails {
-    /// Whether the swap is accepted.
-    /// If Some, contains the tweakable point for this swap.
-    /// If None, swap is rejected.
-    pub tweakable_point: Option<PublicKey>,
+    /// The maker's tweakable public key for this swap.
+    pub tweakable_point: PublicKey,
+    /// Unique session ID for this connection (32 bytes).
+    pub session_id: [u8; 32],
```
Comment (mojoX911) [inline src/protocol/common_messages.rs]: We can self-derive the session i9d and match the signature. No need to pass it around.

File: src/maker/handlers.rs [hunk 10]
```diff
@@ -540,6 +554,7 @@ fn restore_state_if_needed<M: Maker>(maker: &Arc<M>, state: &mut ConnectionState
             state.contract_feerate = stored.contract_feerate;
             state.service_fee_sats = stored.service_fee_sats;
             state.swap_start_time = stored.swap_start_time;
+            state.refund_locktime_offset = stored.refund_locktime_offset
```
Comment (YoganshSharma) [inline src/maker/handlers.rs]: Currently this lines restores `refund_locktime_offset` to zero to make the tests pass. As the connection being persistent accidentally fixes #946.

**Comment by mojoX911** [comment]
@YoganshSharma can you pls rebase and squash the commits?

File: src/utill.rs [hunk 11]
```diff
@@ -981,6 +983,96 @@ pub fn interactive_select(
...
+    Maker,
+    Taker {
+        /// True iff the maker's `session_id_sig` was verified against its tweakable key during the handshake.
+        /// This field is only relevant to Taker
+        is_authenticated: bool,
+    },
+}
+
+pub(crate) struct Bip324Stream {
+    pub protocol: bip324::io::Protocol<std::io::BufReader<TcpStream>, TcpStream>,
```
Comment (mojoX911) [inline src/utill.rs]: Protocol is a confusing name here. Call it `stream`.

File: src/utill.rs [hunk 12]
```diff
@@ -981,6 +983,96 @@ pub fn interactive_select(
...
+    }
+
+    /// Reads the next genuine application message from the BIP324 stream.
+    pub(crate) fn read_message<T: serde::de::DeserializeOwned>(&mut self) -> Result<T, NetError> {
+        loop {
+            let payload = self.protocol.read()?;
+            match payload.packet_type() {
+                // currently we never send decoys but we are handling it
+                bip324::PacketType::Decoy => {
+                    log::debug!("Skipped decoy packet");
```
Comment (mojoX911) [inline src/utill.rs]: Silently skip for now. debug log not needed here.  We can add logging in future if we decide to use decoys.

File: src/utill.rs [hunk 13]
```diff
@@ -981,6 +983,96 @@ pub fn interactive_select(
...
+pub(crate) enum CoinswapRole {
+    Maker,
+    Taker {
+        /// True iff the maker's `session_id_sig` was verified against its tweakable key during the handshake.
+        /// This field is only relevant to Taker
+        is_authenticated: bool,
+    },
+}
+
+pub(crate) struct Bip324Stream {
```
Comment (mojoX911) [inline src/utill.rs:993]: Even though pub(crate) add few lines of good docs for new structs and functions. Applies in other places too.

Comment (mojoX911) [inline src/utill.rs:993]: Structuring: extract all BIP324 stuff into a new module at the repo root: `bip324.md`. This logic are better placed in their own dedicated module. not in util.rs.

File: Cargo.toml [hunk 14]
```diff
@@ -39,6 +39,7 @@ rustls = { version = "0.23.40", default-features = false, features = ["std", "ri
 nostr = "0.44.2"
 crossbeam-channel = "0.5.15"
 hotpath = "0.15.0"
+bip324 = { git = "https://github.com/rust-bitcoin/bip324", rev = "7c8894452929e1f1019e27a84ef8d6834facaa47" } # TODO move to v0.12.0 when it releases
```
Comment (mojoX911) [inline Cargo.toml:41]: Any ETA when will we get this in a release?

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Skimmed through it from structural perpective. Will look deeper into the internals soon. 

What would help in review is in the code sites where major new business logic are added or removed, add an inline code comment there. Its will be helpful for ture reference too. Just explain in 1/2 lines why this is added.
